### PR TITLE
[CDAP-15306] Make Metadata-related mutations from CDAP master synchronous

### PR DESCRIPTION
**Pull request by kmeelu**
Made a new PR for this issue, as last one was corrupted by master merge. 

JIRA: https://issues.cask.co/browse/CDAP-15306
RUT: https://builds.cask.co/browse/CDAP-RUT1672-7
DBV: https://builds.cask.co/browse/CDAP-DBV37-2
ITM: https://builds.cask.co/browse/IT-ITM-232

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java [hunk 1]
```diff
@@ -447,8 +447,7 @@ public void deleteArtifact(Id.Artifact artifactId) throws Exception {
     // delete the artifact first and then privileges. Not the other way to avoid orphan artifact
     // which does not have any privilege if the artifact delete from store fails. see CDAP-6648
     artifactStore.delete(artifactId);
-    metadataPublisher.publish(NamespaceId.SYSTEM,
-                              new MetadataOperation.Drop(artifactId.toEntityId().toMetadataEntity()));
+    metadataServiceClient.remove(new MetadataMutation.Remove(artifactId.toEntityId().toMetadataEntity()));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java]: why not drop?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataAdmin.java [hunk 2]
```diff
@@ -145,6 +146,13 @@ Metadata getMetadata(MetadataEntity entity, @Nullable MetadataScope scope, @Null
    */
   void removeTags(MetadataEntity metadataEntity, Set<String> tags, MutationOptions options) throws IOException;
 
+  /**
+   * Applies a raw metadata mutation.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataAdmin.java]: what does raw mean?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java [hunk 3]
```diff
@@ -148,6 +149,7 @@ public static synchronized Injector getInjector(CConfiguration conf, @Nullable S
         ((Service) messagingService).startAndWait();
       }
 
+      injector.getInstance(MetadataService.class).startAndWait();
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java]: It would be better to start transactions first, because in one implementation, MetadataStorage (and hence MetadataService) relies on transaction.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java [hunk 4]
```diff
@@ -89,7 +89,7 @@ public void testWorkflowTags() throws Exception {
     ArtifactId artifactId = NamespaceId.DEFAULT.artifact(appId.getApplication(), "1.0");
     ApplicationWithPrograms appWithPrograms = createAppWithWorkflow(artifactId, appId, workflowName);
     WorkflowSpecification workflowSpec = appWithPrograms.getSpecification().getWorkflows().get(workflowName);
-    SystemMetadataWriterStage systemMetadataWriterStage = new SystemMetadataWriterStage(metadataPublisher);
+    SystemMetadataWriterStage systemMetadataWriterStage = new SystemMetadataWriterStage(metadataServiceClient);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java:90]: I a wondering whether we should now also change the test to expect the metadata changes instantly instead of waiting for the change to show up with Tasks.waitFor(). In line 98 and 108, perhaps more places?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java [hunk 5]
```diff
@@ -124,6 +126,7 @@ protected void configure() {
                                         bind(UGIProvider.class).to(UnsupportedUGIProvider.class);
                                         bind(OwnerAdmin.class).to(NoOpOwnerAdmin.class);
                                         bind(MetadataPublisher.class).to(NoOpMetadataPublisher.class);
+                                        bind(MetadataServiceClient.class).to(NoOpMetadataServiceClient.class);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java:126]: Is MetadataPublisher still needed then?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java [hunk 6]
```diff
@@ -162,7 +162,7 @@ public void testPublish() throws Exception {
         actualAuditEntities.put(message.getType(), entityId);
       }
       return actualAuditEntities;
-    }, 5, TimeUnit.SECONDS);
+    }, 15, TimeUnit.SECONDS);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java]: why the increase? Now that it is synchronous, I would expect that we don't need to wait at all. How come we need to wait longer?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.Map;
+
+public class DefaultMetadataServiceClientTest extends AppFabricTestBase {
+  // - keep description if new metadata does not contain it
+  // - preserve creation-time if it exists in current metadata
+  private static final Map<ScopedNameOfKind, MetadataDirective> CREATE_DIRECTIVES = ImmutableMap.of(
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:39]: why is this needed if you don't create any metadata of these names?

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    createMetadataMutation(new MetadataMutation.Create(testEntity, toCreate, CREATE_DIRECTIVES));
+
+    Assert.assertEquals(toCreate.getProperties(MetadataScope.SYSTEM),
+                        getMetadataProperties(testEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(toCreate.getTags(MetadataScope.SYSTEM),
+                        getMetadataTags(testEntity, MetadataScope.SYSTEM));
+    Assert.assertEquals(toCreate.getProperties(MetadataScope.USER),
+                        getMetadataProperties(testEntity, MetadataScope.USER));
+    Assert.assertEquals(toCreate.getTags(MetadataScope.USER),
+                        getMetadataTags(testEntity, MetadataScope.USER));
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java]: this should test that if you create again, the properties to-keep and to-preserve are not changed.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  }
+
+  @Test
+  public void testRemove() throws Exception {
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:85]: do these tests rely on being executed in a certain order? Ideally every test uses a different entity and is self-contained.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataCodecTest.java [hunk 10]
```diff
@@ -22,6 +22,7 @@
 import com.google.gson.GsonBuilder;
 import io.cdap.cdap.api.metadata.MetadataEntity;
 import io.cdap.cdap.spi.metadata.Metadata;
+import io.cdap.cdap.spi.metadata.MetadataCodec;
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataCodecTest.java]: why needed? if there is no other code change?

File: cdap-data-fabric/pom.xml [hunk 11]
```diff
@@ -80,7 +80,7 @@
     <dependency>
       <groupId>io.cdap.cdap</groupId>
       <artifactId>cdap-elastic</artifactId>
-      <version>6.0.0-SNAPSHOT</version>
+      <version>${project.version}</version>
```
Comment (anew) [inline cdap-data-fabric/pom.xml:83]: good catch

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java [hunk 12]
```diff
@@ -570,7 +570,7 @@ private void dropDataset(DatasetId instance, DatasetSpecification spec) throws E
 
     // Remove metadata for the dataset
     LOG.trace("Removing metadata for dataset {}", instance);
-    metadataPublisher.publish(NamespaceId.SYSTEM, new MetadataOperation.Drop(instance.toMetadataEntity()));
+    metadataServiceClient.remove(new MetadataMutation.Remove(instance.toMetadataEntity()));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: why not Drop?

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/DefaultMetadataServiceClient.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+  }
+
+  @Override
+  public void create(MetadataMutation.Create createMutation) {
+    HttpRequest request = remoteClient.requestBuilder(HttpMethod.POST, "metadata-internals/create")
+      .withBody(GSON.toJson(createMutation)).build();
+    HttpResponse response = execute(request);
+
+    if (HttpResponseStatus.OK.code() != response.getResponseCode()) {
+      LOG.trace("Failed to create metadata, i.e. see response for details: %s", response);
```
Comment (anew) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/DefaultMetadataServiceClient.java]: Can it just be "Failed to create metadata: < response>". Also, does the response contain the entity id? If not, might be good to change it to: "Failed to create metadata for entity < ...>: < response>"

Comment (anew) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/DefaultMetadataServiceClient.java]: and same for the other methods

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+import io.cdap.cdap.spi.metadata.MetadataMutation;
+
+/**
+ * This interface exposes functionality for making Metadata HTTP calls to the Metadata Service.
+ */
+public interface MetadataServiceClient {
+
+  /**
+   * Publishes the create mutation to Metadata Service.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java]: Can you rephrase this? "Publish" does not sound as if it is a synchronous call.

File: cdap-gateway/src/test/java/io/cdap/cdap/gateway/GatewayTestBase.java [hunk 15]
```diff
@@ -221,6 +225,7 @@ public static void stopGateway(CConfiguration conf) throws Exception {
     datasetService.stopAndWait();
     dsOpService.stopAndWait();
     txService.stopAndWait();
+    metadataService.stopAndWait();
```
Comment (anew) [inline cdap-gateway/src/test/java/io/cdap/cdap/gateway/GatewayTestBase.java:227]: it is best practice to stop all services in reverse order of starting.

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java [hunk 16]
```diff
@@ -98,6 +100,7 @@ protected void configure() {
           bind(OwnerAdmin.class).to(DefaultOwnerAdmin.class);
           // TODO (CDAP-14677): find a better way to inject metadata publisher
           bind(MetadataPublisher.class).to(MessagingMetadataPublisher.class);
+          bind(MetadataServiceClient.class).to(DefaultMetadataServiceClient.class);
```
Comment (anew) [inline cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java]: Does MetadataService itself depend on the metadata service client? Does it make HTTP calls to itself? I hope not, and then we should not need this?

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/ScopedNameOfKind.java [hunk 17]
```diff
@@ -66,4 +66,10 @@ public int hashCode() {
   public String toString() {
     return '(' + kind.name().toLowerCase() + ')' + super.toString();
   }
+
+  public static ScopedNameOfKind fromString(String s) {
```
Comment (anew) [inline cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/ScopedNameOfKind.java:70]: what will this method do if it is not a well-formed ScopedNameOfKind? It should throw IllegalArgumentException with a meaningful message, including the string that is being parsed. I don't think this method does that? It would throw ArrayOutOfBound exception or similar.

**Review by anew — COMMENTED** [review]
Overall looks very good. Please see the comments I had. 

Also: 
- are there more places where we can now remove the Tasks.waitFor(), because metadata is written synchronously?
- Can we simplify the metadata operations (and the subscriber)? Or are all kinds of messages still possible? I feel that the Create and Drop probably are never published any more.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java (hunk 1 above)
Comment (kmeelu) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/artifact/DefaultArtifactRepository.java]: good point. Not sure why I used this instead. Fixing now.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataAdmin.java (hunk 2 above)
Comment (kmeelu) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/metadata/MetadataAdmin.java]: Updating phrasing to "applies a metadata mutation directly". The point here is that while the other functions create the MetadataMutation to apply, this simply accepts the mutation itself and applies it. I suppose "raw" doesn't describe that.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java (hunk 6 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java]: This was failing earlier, but I didn't intend for it to come through in the PR. Removing; since it's a publish test, the waitFor is waiting to fetch asynchronous messages. It seems to use TMS to send non-metadata messages, so I won't modify it.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataCodecTest.java (hunk 10 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/MetadataCodecTest.java]: Oops, intellij added this since I moved the MetadataCodec class to cdap-metadata-spi (for access from cdap-data-fabric) but didn't move the test. Moving this to cdap-metadata-spi tests, as well.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java (hunk 12 above)
Comment (kmeelu) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/datafabric/dataset/service/DatasetInstanceService.java]: Resolving this, as well. Not sure why I made this a remove.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java (hunk 3 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java]: Ok, done. In TestBase, GatewayTestBase, and here, I'll order it as Transactions, MetadataStorage, then MetadataServiceClient (and reverse for stopping).

File: cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/ScopedNameOfKind.java (hunk 17 above)
Comment (kmeelu) [reply cdap-metadata-spi/src/main/java/io/cdap/cdap/spi/metadata/ScopedNameOfKind.java:70]: Using ScopedName.fromString() as a template, I'm adding a check to make sure that opening parenthesis is in first position of string and the closing parenthesis is in index >1.

File: cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java (hunk 16 above)
Comment (kmeelu) [reply cdap-master/src/main/java/io/cdap/cdap/master/environment/k8s/MetadataServiceMain.java]: fair point. May have gotten trigger happy adding this. removing.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java (hunk 14 above)
Comment (kmeelu) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java]: Performs the mutation via Metadata Service?

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/DefaultMetadataServiceClient.java (hunk 13 above)
Comment (kmeelu) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/DefaultMetadataServiceClient.java]: Done.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java (hunk 9 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:85]: Yes, fixing this.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java (hunk 8 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java]: Ok, going hand-in-hand with isolating tests from each other, I'm creating the entity in each. However, testing Keep and Preserve in the Remove and Update tests.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java (hunk 7 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:39]: Good point, updating these, based on our conversation.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java (hunk 4 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/deploy/pipeline/SystemMetadataWriterStageTest.java:90]: Updated these, and will do so for other tests in this scope.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java (hunk 5 above)
Comment (kmeelu) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/app/runtime/schedule/store/DatasetBasedTimeScheduleStoreTest.java:126]: Don't think so. Removing it wherever NoOpMetadataServiceClient has been added.

**Review by kmeelu — COMMENTED** [review]
Made review changes. PTAL

**Comment by kmeelu** [comment]
> Overall looks very good. Please see the comments I had.
> 
> Also:
> 
> * are there more places where we can now remove the Tasks.waitFor(), because metadata is written synchronously?
> * Can we simplify the metadata operations (and the subscriber)? Or are all kinds of messages still possible? I feel that the Create and Drop probably are never published any more.

Regarding these overarching questions, 
1) I removed the Tasks.waitFor() in the places I found in this diff (only ~2 files). However, because AppFabricTestBase surfaces metadataStorage, it's possible tests that extend this would have Tasks.waitFor() that could be removed. This could be part of this PR or another issue; let me know how you'd like to deal with this.
2) I haven't looked into this, but all four of the operations do get surfaced to a runtime class. I can try digging that up; I'm unsure where/whether those runtime functions get called.

File: cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java (hunk 14 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/writer/MetadataServiceClient.java]: sure that works.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java (hunk 3 above)
Comment (anew) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/AppFabricTestHelper.java]: +1

File: cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java (hunk 6 above)
Comment (anew) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/internal/audit/AuditPublishTest.java]: ok

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+                        getMetadataProperties(createEntity, MetadataScope.USER));
+    Assert.assertEquals(testMetadata.getTags(MetadataScope.USER),
+                        getMetadataTags(createEntity, MetadataScope.USER));
+  }
+
+  @Test
+  public void testRemove() throws Exception {
+    final MetadataEntity removeEntity = MetadataEntity.builder().append("remove", "test").build();
+    createMetadataMutation(new MetadataMutation.Create(removeEntity, testMetadata, CREATE_DIRECTIVES));
+
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:88]: Please assert that the create was successful.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+                        getMetadataProperties(updateEntity, MetadataScope.USER));
+    Assert.assertEquals(testMetadata.getTags(MetadataScope.USER),
+                        getMetadataTags(updateEntity, MetadataScope.USER));
+  }
+
+  @Test
+  public void testDrop() throws Exception {
+    final MetadataEntity dropEntity = MetadataEntity.builder().append("drop", "test").build();
+    createMetadataMutation(new MetadataMutation.Create(dropEntity, testMetadata, CREATE_DIRECTIVES));
+
```
Comment (anew) [inline cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:158]: assert that the metadata was actually created.

File: cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java (hunk 18 above)
Comment (anew) [reply cdap-app-fabric/src/test/java/io/cdap/cdap/metadata/DefaultMetadataServiceClientTest.java:88]: Can you do a remove that only removes specific tags/props? And then validate that only those were removed.

**Review by anew — COMMENTED** [review]
Everything look good now. Except a couple of comments for the client tests. Good work.

**Review by anew — APPROVED** [review]
LGTM if build passes

**Comment by anew** [comment]
closed this, instead we have #11349  against develop
